src/PaletteProvider.js

```diff
@@ -144,35 +126,47 @@ export default class MaterialPaletteProvider
     if (this.props.defaults) {
       validateDefaults(this.props.defaults);
     }
-    execIfFunction(this.props.onInit);
-    createMaterialPalette(this.props.image, this.props.options)
-      .then((palette: PaletteInstance) => {
+
+    this._createPalette();
+  }
+
+  _createPalette = () => {
+    const { image, options, onFinish, onError } = this.props;
+
+    createMaterialPalette(image, options).then(
+      palette => {
         const paletteWithDefaults = this._mergeWithDefaults(palette);
-        execIfFunction(this.props.onFinish, paletteWithDefaults);
-        if (!this.props.forceRender) {
-          this.setState({ palette: paletteWithDefaults });
-        }
+
+        if (onFinish) onFinish(paletteWithDefaults);
+
+        this.setState({ palette: paletteWithDefaults });
         this.eventEmitter.publish({
           palette: paletteWithDefaults,
         });
-      })
-      .catch((error: Error) => {
-        const isCalled = execIfFunction(this.props.onError, error);
-        if (!isCalled) {
-          const enhancedError = error;
-          enhancedError.message = `Uncaught MaterialPaletteProvider exception: ${enhancedError.message}`;
-          throw enhancedError;
+      },
+      error => {
+        if (onError) {
+          onError(error);
+        } else {
+          error.message = `Uncaught MaterialPaletteProvider exception: ${error.message}`; // eslint-disable-line no-param-reassign
+          throw error;
         }
-      });
-  }
+      },
+    );
+  };
 
   render() {
-    if (!this.state.palette && this.props.LoaderComponent) {
-      return <this.props.LoaderComponent />;
-    } else if (!this.state.palette && !this.props.forceRender) {
-      return null;
+    const { forceRender, LoaderComponent, children } = this.props;
+    const shouldRender = this.state.palette || forceRender;
+
+    if (LoaderComponent && !this.state.palette) {
+      return <LoaderComponent />;
+    }
+
+    if (shouldRender) {
```

Comment:
shouldn't we change the order? If `forceRender` is provided, it seems natural for me to override `LoaderComponent`, what do you think?